runtime/auth/aws-signing-common/common/test/aws/smithy/kotlin/runtime/auth/awssigning/AwsChunkedByteReadChannelTest.kt

```diff
@@ -0,0 +1,913 @@
+        override suspend fun getCredentials(): Credentials = this@asStaticProvider
+    }
+
+    private val testSigner = DefaultAwsSigner
+    private val testSigningConfig = AwsSigningConfig {
+        region = "foo"
+        service = "bar"
+        signingDate = Instant.fromIso8601("20220427T012345Z")
+        credentialsProvider = testCredentialsProvider
+    }
+
+    /**
+     * Given a string representation of aws-chunked encoded bytes, return a list of the chunk signatures as strings.
+     * Chunk signatures are defined by the following grammar:
+     * chunk-signature=<64 alphanumeric characters>
+     */
+    private fun getChunkSignatures(bytes: String): List<String> = CHUNK_SIGNATURE_REGEX.findAll(bytes).map {
+            result ->
+        result.value.split("=")[1]
+    }.toList()
+
+    /**
+     * Given a string representation of aws-chunked encoded bytes, returns a list of the chunk sizes as integers.
+     * Chunk sizes are defined by the following grammar:
+     * String(Hex(ChunkSize));chunk-signature=<chunk_signature>
+     */
+    private fun getChunkSizes(bytes: String): List<Int> = CHUNK_SIZE_REGEX.findAll(bytes).map {
+            result ->
+        result.value.split(";")[0].toInt(16)
+    }.toList()
+
+    /**
+     * Given a string representation of aws-chunked encoded bytes, return the value of the x-amz-chunk-trailer trailing header.
+     */
+    private fun getChunkTrailerSignature(bytes: String): String? {
+        val re = Regex("x-amz-trailer-signature:[a-zA-Z0-9]{64}")
+        return re.findAll(bytes).map { result ->
+            result.value.split(":")[1]
+        }.toList().firstOrNull()
+    }
+
+    /**
+     * Calculates the `aws-chunked` encoded trailing header length
+     * Used to calculate how many bytes should be read for all the trailing headers to be consumed
+     */
+    private fun getTrailingHeadersLength(trailingHeaders: Headers) = trailingHeaders.entries().map {
+            entry ->
+        buildString {
+            append(entry.key)
+            append(":")
+            append(entry.value.joinToString(","))
+            append("\r\n")
+        }.length
+    }.reduce { acc, len -> acc + len } +
+        "x-amz-trailer-signature:".length + 64 + "\r\n".length
+
+    @Test
+    fun testReadRemainingExactBytes() = runTest {
+        val dataLengthBytes = CHUNK_SIZE_BYTES
+        val data = ByteArray(dataLengthBytes) { 0x7A.toByte() }
+        val chan = SdkByteReadChannel(data)
+        val previousSignature: ByteArray = byteArrayOf()
+        val awsChunked = AwsChunkedByteReadChannel(chan, testSigner, testSigningConfig, previousSignature)
+
+        // read the exact chunk length
+        val bytes = awsChunked.readRemaining(
+            CHUNK_SIZE_BYTES + CHUNK_SIZE_BYTES.toString(16).length + 1 +
+                "chunk-signature=".length + 64 + 4 + 1 + 1 + "chunk-signature=".length + 64 + 4 + 2,
```

Comment:
This seems like a bunch of magical numbers...can we get some comments in here or define some constants to make the test code clearer? (this goes for several tests, not going to call them all out - the tests are good but they kind of assume an intimate knowledge that hinders overall test understanding/maintenance)